Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,236 @@
+import numpy as np
+import numbers
+
+
+class SVRGFunction(ApproximateGradientSumFunction):
+
+    """
+    A class representing a function for Stochastic Variance Reduced Gradient (SVRG) approximation. 
+
+    Parameters
+    ----------
+    functions : list
+        A list of functions to optimize.
+    sampler : callable or None, optional
+         A callable function to select the next function, see e.g. optimisation.utilities.sampler 
+    update_frequency : int or None, optional
+        The frequency of updating the full gradient. The default is 2*len(functions)
+    store_gradients : bool, default: `False`
+        Flag indicating whether to store gradients for each function.
+
+    Reference
+    ---------
+    Johnson, R. and Zhang, T., 2013. Accelerating stochastic gradient descent using predictive variance reduction. Advances in neural information processing systems, 26.
+    
+    """
+
+    def __init__(self, functions, sampler=None, update_frequency=None, store_gradients=False):
+        super(SVRGFunction, self).__init__(functions, sampler)
+
+        # update_frequency for SVRG
+        self.update_frequency = update_frequency
+    
+        if self.update_frequency is None:
+            self.update_frequency = 2*self.num_functions
+
+        # compute and store the gradient of each function in the finite sum
+        self.store_gradients = store_gradients
+
+        self._svrg_iter_number = 0
+
+    def gradient(self, x, out=None):
+        """ Selects a random function using the `sampler` and then calls the approximate gradient at :code:`x` or calculates a full gradient depending on the update frequency
+
+        Parameters
+        ----------
+        x : DataContainer
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+
+        Returns
+        --------
+        DataContainer
+            the value of the approximate gradient of the sum function at :code:`x` 
+        """
+
+
+    
+
+        if (np.isinf(self.update_frequency) == False and (self._svrg_iter_number % (self.update_frequency)) == 0):
+
+            return self._update_full_gradient_and_return(x, out=out)
+
+        else:
+
+            self.function_num = self.sampler.next()
+
+            if not isinstance(self.function_num, numbers.Number):
+                raise ValueError("Batch gradient is not yet implemented")
+            if self.function_num > self.num_functions:
+                raise IndexError(
+                    'The sampler has outputted an index larger than the number of functions to sample from. Please ensure your sampler samples from {0,1,...,len(functions)-1} only.')
+
+            return self.approximate_gradient(x, self.function_num, out=out)
+
+        
+
+    def approximate_gradient(self, x, function_num, out=None):
+        """ Computes the approximate gradient for each selected function at :code:`x` given a `function_number` in {0,...,len(functions)-1}.
```

Comment:
PLease check it is explained in what way the gradient is being approximated in the SVRG method. Similar to how it was described for SG that it is the gradient of a single term in the sum.